soc: arm: Add Aspeed ast10x0 series support

aspeeddylan opened the pull request:
Aspeed ast10x0 series SoCs contain a ARM Cortex-M4F processor.  This
processor operates at 200MHz and executes on SRAM.

This patch adds support for ast1030 as the first SoC of this series
which is targeted but not limited at the bridge IC in a server system.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
Change-Id: I668af1ff8a36a05da791c3329ae08f5ae712bdd4

aspeeddylan committed:
soc: arm: Add Aspeed ast10x0 series support

Aspeed ast10x0 series SoCs contain a ARM Cortex-M4F processor.  This
processor operates at 200MHz and executes on SRAM.

This patch adds support for ast1030 as the first SoC of this series
which is targeted but not limited at the bridge IC in a server system.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
Change-Id: I668af1ff8a36a05da791c3329ae08f5ae712bdd4

aspeeddylan commented:
The problem I am encountered here is the same with [PR789](https://github.com/zephyrproject-rtos/zephyr/pull/789/commits/c78c95b4ce86794d4be8293aa00fa17332971c40)

In my local branch, I have a device with IRQ number greater than 128 so I got failed to enable its IRQ.  I found the solution used in PR789 also works for me, but now I have problem on upstreaming this part of code.  Could anyone advise me how to fix this? Thank you.

```
-:192: WARNING:NEW_TYPEDEFS: do not add new typedefs
24
#192: FILE: soc/arm/aspeed/aspeed_util.h:59:
25
+typedef enum {
26

27
- total: 0 errors, 1 warnings, 520 lines checked
```

MaureenHelm commented on soc/arm/aspeed/aspeed_util.h:
```diff
@@ -0,0 +1,86 @@
+#define ALIGNED16_SECTION(name)         (aligned(16), section(name))
+#define __section_aligned16(name)       __attribute__(ALIGNED16_SECTION(name))
+
+/* non-cached (DMA) memory */
+#if (CONFIG_SRAM_NC_SIZE > 0)
+#define NON_CACHED_BSS                  __section(".nocache.bss")
+#define NON_CACHED_BSS_ALIGN16          __section_aligned16(".nocache.bss")
+#else
+#define NON_CACHED_BSS
+#define NON_CACHED_BSS_ALIGN16          __aligned(16)
+#endif
```
You can set `CONFIG_NOCACHE_MEMORY` instead

aspeeddylan replied:
> This SoC needs a board definition to go with it (preferably a second commit added to the PR); otherwise none of this code will get built in CI.

Got it.  I am preparing the second commit and will add it to this PR once I finish it.

>you can set CONFIG_NOCACHE_MEMORY instead

`CONFIG_NOCACHE_MEMORY` depends on `CONFIG_ARCH_HAS_NOCACHE_MEMORY_SUPPORT` which is only enabled on Cortex-M7 for ARM Cortex-M processors.  The CPU in AST10X0 series SOC is Cortex-M3 so I couldn't set  `CONFIG_NOCACHE_MEMORY` for our chip.  Could you kindly advice me if there is a better way to handle this?

aspeeddylan replied:
Force push to add second commit for the ast1030 evaluation board support

aspeeddylan replied:
@MaureenHelm,
Should I resolve the conversation? Or would kindly advice me?

MaureenHelm reviewed (dismissed):
Thank you for your contribution!

This SoC needs a board definition to go with it (preferably a second commit added to the PR); otherwise none of this code will get built in CI.

aspeeddylan committed:
boards: arm: Add ast1030_evb board

Skeleton board support for the aspeed ast1030 evaluation board from
Aspeed Technology.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
Change-Id: I16af50dd63cb355071e32f89264d8aeae47184ac

aspeeddylan commented:
force push to fix the ```new typedef``` warning.  Use ```IRQn_Type``` defined in ```cmsis.h```

GoldenBug commented:
Hi Zephyr team, @MaureenHelm 

Are there any blockers for reviewing this or any other changes requested?

cfriedt commented on soc/arm/aspeed/ast10x0/tools/gen_uart_booting_image.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
```
I would recommend rewriting this script in Python (v3), not because I'm personally against bash (or whatever shell happens to be installed on a given system), but because Python is a more consistent scripting environment that works across all platforms supported by Zephyr.

Is that doable?

aspeeddylan replied:
Hi @cfriedt ,

Yes, it is doable.  And just now I edit my first commit to replace the shell script by the Python based tool.  Please kindly review my change, thank you very much.

cfriedt commented:
@aspeeddylan - LGTM other than the one issue I mentioned. I'll ack once the shell script is converted to Python.

aspeeddylan commented:
> @aspeeddylan - LGTM other than the one issue I mentioned. I'll ack once the shell script is converted to Python.

Hi @cfriedt,
Thank you for your review. I have replaced the shell script by the python based tool in my first commit.

cfriedt reviewed (dismissed):
@aspeeddylan - it looks like `build docs` is erroring, otherwise LGTM.

aspeeddylan commented:
> @aspeeddylan - it looks like `build docs` is erroring, otherwise LGTM.

Not sure why `build docs` failed since it was successfully built.  Should I fetch and merge to the upstream?

GoldenBug commented:
> > @aspeeddylan - it looks like `build docs` is erroring, otherwise LGTM.
> 
> Not sure why `build docs` failed since it was successfully built. Should I fetch and merge to the upstream?

Hi Dylan, Can you try rebasing and see if the docs are able to build.
Might have just been a transient issue.

aspeeddylan commented:
> > > @aspeeddylan - it looks like `build docs` is erroring, otherwise LGTM.
> > 
> > 
> > Not sure why `build docs` failed since it was successfully built. Should I fetch and merge to the upstream?
> 
> Hi Dylan, Can you try rebasing and see if the docs are able to build. Might have just been a transient issue.

Hi @GoldenBug,
Thank you for your comments. It's OK now.

aspeeddylan commented:
@cfriedt I re-request for your review by mistake, could you review my PR again, thank you very much.

GoldenBug commented:
Is there anyone who can complete a second review of this?
@cfriedt

carlescufi reviewed (approved):
LGTM
